enclave/core/debugmalloc.c

```diff
@@ -474,4 +487,270 @@ size_t oe_debug_malloc_check(void)
+    // that posix_memalign requires alignment to be a multiple of sizeof(void*).
+    // Adjust the alignment if needed.
+    alignment = oe_round_up_to_multiple(alignment, sizeof(void*));
+
+    oe_posix_memalign(&ptr, alignment, size);
+    return ptr;
+}
+
+int oe_posix_memalign(void** memptr, size_t alignment, size_t size)
+{
+    int rc = -1;
+
+    if (oe_use_debug_malloc)
+        rc = oe_debug_posix_memalign(memptr, alignment, size);
+    else
+        rc = oe_posix_memalign(memptr, alignment, size);
+
+    if (rc != 0 && size)
+    {
+        if (_failure_callback)
+            _failure_callback(__FILE__, __LINE__, __FUNCTION__, size);
+    }
+
+    return rc;
+}
+
+size_t oe_malloc_usable_size(void* ptr)
+{
+    if (oe_use_debug_malloc)
+    {
+        return oe_debug_malloc_usable_size(ptr);
+    }
+    else
+    {
+        return oe_allocator_malloc_usable_size(ptr);
+    }
+}
+
+oe_result_t oe_check_memory_leaks(void)
+{
+    if (!oe_disable_debug_malloc_check && oe_debug_malloc_check() != 0)
+        return OE_MEMORY_LEAK;
+    return OE_OK;
+}
+
+oe_result_t oe_debug_malloc_tracking_start(void)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    oe_spin_lock(&_spin);
+    if (!oe_use_debug_malloc_tracking)
+    {
+        oe_use_debug_malloc_tracking = true;
+        ++oe_debug_malloc_session_number;
+        result = OE_OK;
+    }
+    oe_spin_unlock(&_spin);
+
+    return result;
+}
+
+oe_result_t oe_debug_malloc_tracking_stop(void)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    oe_spin_lock(&_spin);
+    if (oe_use_debug_malloc_tracking)
+    {
+        oe_use_debug_malloc_tracking = false;
+        result = OE_OK;
+    }
+    oe_spin_unlock(&_spin);
+
+    return result;
+}
+
+static oe_result_t _copy_frames(
+    header_t* p,
+    char** str,
+    size_t* size,
+    size_t* index)
+{
+    oe_result_t result = OE_FAILURE;
+    char** syms = NULL;
+
+    if (!(syms = oe_backtrace_symbols(p->addrs, (int)(p->num_addrs))))
+    {
+        goto done;
+    }
+
+    for (uint64_t i = 0; i < p->num_addrs; i++)
+    {
+        size_t length_s = oe_strlen(syms[i]);
+        size_t length_a = sizeof(p->addrs[i]) * 2;
+        size_t length = length_s + length_a + 6;
+
+        if (*index + length >= *size)
+        {
+            while (*index + length >= *size)
+            {
+                *size *= 2;
+            }
+
+            *str = oe_realloc(*str, *size);
+            if (*str == NULL)
+            {
+                result = OE_ENOMEM;
+                goto done;
+            }
+        }
+
+        oe_snprintf(*str + *index, length, "%s(): %p\n", syms[i], p->addrs[i]);
+        *index = oe_strlen(*str);
+    }
+
+    (*str)[(*index)++] = '\n';
+    (*str)[*index] = '\0';
+
+    result = OE_OK;
+
+done:
+    oe_backtrace_symbols_free(syms);
+
+    return result;
+}
+
+oe_result_t oe_debug_malloc_tracking_report(
+    uint64_t* out_object_count,
+    char** report)
```

Comment:
Now that we track the session number along with the object, we have the ability to
- generate a tracking report for last session
   This allows finding out objects that were allocated in the last session but are still alive.
- generate a tracking report for a given session
    This allows finding out objects that were allocated in a given session but are still alive
- generate an aggregated tracking report (current behavior)
   This allows finding out objects that are still alive.

Generating reports based on session numbers would be something we could do in the future.